apk/apk.go

```diff
@@ -399,54 +399,70 @@ func createBuilderData(info *nfpm.Info, sizep *int64) func(tw *tar.Writer) error
 	created := map[string]bool{}
 
 	return func(tw *tar.Writer) error {
-		// handle empty folders
-		if err := createEmptyFoldersInsideTarGz(info, tw, created); err != nil {
-			return err
-		}
-
-		// handle Files
 		return createFilesInsideTarGz(info, tw, created, sizep)
 	}
 }
 
 func createFilesInsideTarGz(info *nfpm.Info, tw *tar.Writer, created map[string]bool, sizep *int64) (err error) {
+	// create explicit directories first
+	for _, file := range info.Contents {
+		// at this point, we don't are about other types yet
+		if file.Type != "dir" {
+			continue
+		}
+
+		// only consider contents for this packager
+		if file.Packager != "" && file.Packager != packagerName {
+			continue
+		}
+
+		err = tw.WriteHeader(&tar.Header{
+			Name:     file.Destination,
+			Mode:     int64(file.FileInfo.Mode),
+			Typeflag: tar.TypeDir,
+			Format:   tar.FormatGNU,
+			Uname:    file.FileInfo.Owner,
+			Gname:    file.FileInfo.Group,
+			ModTime:  file.FileInfo.MTime,
+		})
+		if err != nil {
+			return err
+		}
+
+		created[strings.TrimPrefix(file.Destination, "/")] = true
+	}
+
 	for _, file := range info.Contents {
+		// only consider contents for this packager
 		if file.Packager != "" && file.Packager != packagerName {
 			continue
 		}
+
+		// create implicit directory structure below the current content
 		if err = createTree(tw, file.Destination, created); err != nil {
 			return err
 		}
+
 		switch file.Type {
 		case "ghost":
 			// skip ghost files in apk
 			continue
+		case "dir":
+			// already handled above
+			continue
 		case "symlink":
 			err = createSymlinkInsideTarGz(file, tw)
-		case "doc":
-			// nolint:gocritic
-			// ignoring `emptyFallthrough: remove empty case containing only fallthrough to default case`
-			fallthrough
-		case "licence", "license":
-			// nolint:gocritic
-			// ignoring `emptyFallthrough: remove empty case containing only fallthrough to default case`
-			fallthrough
-		case "readme":
+		case "doc", "licence", "license", "readme", "config", "config|noreplace":
 			// nolint:gocritic
-			// ignoring `emptyFallthrough: remove empty case containing only fallthrough to default case`
-			fallthrough
-		case "config", "config|noreplace":
-			// nolint:gocritic
-			// ignoring `emptyFallthrough: remove empty case containing only fallthrough to default case`
+			// ignoring `emptyFallthrough: remove empty case containing only
+			// fallthrough to default case`
 			fallthrough
 		default:
 			err = copyToTarAndDigest(file, tw, sizep)
 		}
 		if err != nil {
 			return err
 		}
-		created[file.Source] = true
-		created[file.Destination[1:]] = true
```

Comment:
I don't know why these two lines where there to begin with. The source only matters for symlinks but created is only to avoid creating duplicate implicit directories anyway, right?

Reply:
I don't know either 🤔  but I think if tests are passing its probably something we didn't need indeed..